src/BinlogTool/DumpRecords.cs

```diff
@@ -0,0 +1,94 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Build.Logging.StructuredLogger;
+using Microsoft.Build.Logging;
+using StructuredLogger.Utils;
+
+namespace BinlogTool
+{
+    internal static class DumpRecords
+    {
+        public static void Run(List<string> inputs, bool includeTotal, bool includeRollup, bool includeDetails)
+        {
+            if (!inputs.Any())
+            {
+                // Default - current directory
+                inputs.Add(string.Empty);
+            }
+
+            var inputBinlogs = inputs.SelectMany(input => Searcher.FindBinlogs(input, true)).ToList();
+
+            if (!inputBinlogs.Any())
+            {
+                Log.WriteError("No binlogs found.");
+                return;
+            }
+
+            if (inputBinlogs.Count > 1)
+            {
+                Log.WriteLine(
+                    $"Found {inputBinlogs.Count} binlog files. Will provide overview for all. (found files: {(string.Join(',', inputBinlogs))})");
+            }
+
+            foreach (var inputBinlog in inputBinlogs)
+            {
+                if (inputBinlogs.Count > 1)
+                {
+                    Log.WriteLine();
+                    Log.WriteLine($"Overview for {inputBinlog}");
+                    Log.WriteLine();
+                }
+
+                RunSingle(inputBinlog, includeTotal, includeRollup, includeDetails);
+            }
+        }
+
+        public static void RunSingle(string input, bool includeTotal, bool includeRollup, bool includeDetails)
+        {
+            var reader = BinaryLogReplayEventSource.OpenBuildEventsReader(input);

```

Comment:
If we can't have those types public, perhaps add a new public static API on `BinaryLog` that exposes what you want. It already has ReadRecords and friends.